src/coreclr/src/jit/lsra.cpp

```diff
@@ -4435,7 +4435,24 @@ void LinearScan::spillGCRefs(RefPosition* killRefPosition)
         {
             continue;
         }
-        unassignPhysReg(regRecord, assignedInterval->recentRefPosition);
+        bool needsKill = varTypeIsGC(assignedInterval->registerType);
+        if (!needsKill)
+        {
+            // The importer will assign a GC type to the rhs of an assignment if the lhs type is a GC type,
+            // even if the rhs is not. See the CEE_STLOC* case in impImportBlockCode(). As a result,
+            // we can have an integer type lclVar that behaves as a GC type. In that case we need to
+            // kill the local.
```

Comment:
Do we need to kill that local because GC will try to add that address as alive or because its value won't be correct after whatever we are spilling for?
Can we improve that and do not spill such locals?

Reply:
We need to spill (not really "kill") the local because we don't support GC refs in registers for unmanaged calls. I think that we could possibly improve the handling of stack references, because even though the address of stack locals is generally marked as a byref, we don't really have to report it to the GC if we know it's on the stack. That said, this case is rare enough (I don't think it can be done in C#) that I'm not sure it's worth optimizing for.